### PR TITLE
drivers: pwm: add a SAM0 TC based PWM driver

### PR DESCRIPTION
**Pull request by Terezvent**
PWM driver using the SAM0 Timer/Counter (TC) Supports the SAMD21 and SAMD5x series, 8 and 16 bit counter size is supported.

The 8-bit counter operates in Normal PWM (NPWM) mode, it supports pulse width and period values between 0 and 255. It is ideal for applications requiring moderate frequency PWM, however, it is not suitable for high-precision or low-frequency applications.

The 16-bit counter operates in Match PWM (MPWM) mode to generate the PWM signal. This mode sacrifices the timer's CC0 channel in order to achieve pulse width modulation.

### CODE REVIEW

File: drivers/pwm/pwm_sam0_tc.c [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+		regs->COUNT16.CC[0].reg = TC_COUNT16_CC_CC(1);
+
+		regs->COUNT16.CTRLA.bit.ENABLE = 1;
+		wait_synchronization(regs, cfg->counter_size);
+	}
+
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_sam0_driver_api = {
```
[the remark this answers is not in the document]
Comment (Terezvent) [reply drivers/pwm/pwm_sam0_tc.c]: Fixed, thanks!

File: drivers/pwm/Kconfig.sam0_tc [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	default y
+	depends on DT_HAS_ATMEL_SAM0_TC_PWM_ENABLED
+	help
+	  Enable PWM driver for Atmel SAM0 MCUs using the TC timer/counter.
+
+config PWM_TC_INIT_PRIORITY
+	int "Init Priority"
+	default 50
+	help
+		Device driver initialization priority
```
Comment (pdgendt) [inline drivers/pwm/Kconfig.sam0_tc:17]: Nit: two space indent

```suggestion
	  Device driver initialization priority
```

File: drivers/pwm/pwm_sam0_tc.c [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+static DEVICE_API(pwm, pwm_sam0_driver_api) = {
+	.set_cycles = pwm_sam0_set_cycles,
+	.get_cycles_per_sec = pwm_sam0_get_cycles_per_sec,
+};
+
+#ifdef MCLK
+#define PWM_SAM0_INIT_CLOCKS(inst)                                                           \
+	.mclk = (volatile uint32_t *)MCLK_MASK_DT_INT_REG_ADDR(inst),                            \
+	.mclk_mask = BIT(DT_INST_CLOCKS_CELL_BY_NAME(inst, mclk, bit)),                          \
+	.gclk_id = DT_INST_CLOCKS_CELL_BY_NAME(inst, gclk, periph_ch)
```
Comment (pdgendt) [inline drivers/pwm/pwm_sam0_tc.c:192]: Nit:

awkward trailing `\`

**Review by pdgendt — COMMENTED** [review]
Thanks, 2 nit formatting comments

File: drivers/pwm/Kconfig.sam0_tc (hunk 2 above)
Comment (Terezvent) [reply drivers/pwm/Kconfig.sam0_tc]: Done, thanks

**Review by pdgendt — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Comment by nandojve** [comment]
new driver require tests.

**Pull request merged**